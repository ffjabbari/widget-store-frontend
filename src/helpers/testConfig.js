/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'A good remote.',
    discount: '0.15',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    rating: null,
    quantity: 1,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: '',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
    set_default: false,
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        date_added: '2021-01-24T08:00:00.000Z',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 2,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };
};
