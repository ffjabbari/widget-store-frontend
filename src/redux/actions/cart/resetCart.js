/** Reset action creator. */

import { RESET_CART } from '../types';

function resetCart() {
  return { type: RESET_CART };
}

export default resetCart;
