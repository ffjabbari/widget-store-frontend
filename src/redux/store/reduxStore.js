/*
 * Uses redux-persist to persist state.
 * Uses redux-thunk to allow async operations.
 */

import { composeWithDevTools } from 'redux-devtools-extension';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';

// local imports
import root from '../reducers/root';

const persistConfig = {
  key: 'root',
  storage,
  // allow any new state properties to persist state rehydration
  stateReconciler: autoMergeLevel2,
};

const persistedReducer = persistReducer(persistConfig, root);

export const store = createStore(
  persistedReducer,
  composeWithDevTools(applyMiddleware(thunk))
);

export const persistedStore = persistStore(store);

export const testStore = createStore(root);
