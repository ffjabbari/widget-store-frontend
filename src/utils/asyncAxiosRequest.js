/** Asynchronous axios request function. */

import axios from 'axios';

const BASE_URL = process.env.REACT_APP_BASE_URL || 'http://localhost:3001';

/** asyncAxiosRequest()
 *
 * Function to abstract error handling of axios requests.
 *
 * @param {string} url
 * @param {string} method
 * @param {object} data
 * @param {object} params
 * @param {string} baseURL
 */

export const asyncAxiosRequest = async (
  url,
  method,
  data,
  params,
  baseURL = BASE_URL
) => {
  try {
    const resp = await axios({ url, method, data, params, baseURL });
    // console.log(resp);
    return resp;
  } catch (error) {
    return { error };
  }
};
