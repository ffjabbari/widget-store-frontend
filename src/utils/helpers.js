import { Redirect, Link } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * getPathRoot()
 *
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

/**
 * <FramerRedirect to="" />
 *
 * Component to force react-router-dom redirect when framer-motion
 * prevents a redirect from executing.
 *
 * If regular <Redirect /> fails in this environment use this redirect instead.
 *
 * @param {any} props
 *
 * returns: motion.div wrapped redirect component
 */
export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);
export const FramerLink = props => (
  <motion.div exit="undefined">
    <Link {...props} />
  </motion.div>
);

/**
 * animateVariant()
 *
 * Animation function to turn on an animation state then turn it back
 * off after a delay. The state will be set "active" then back to "default".
 *
 * If inital state provided by setterFunct is false this function will skip
 * setting the animation state "active" instead set it to "default".
 *
 * If a callback is provided it will be called with any arguments
 * after setting state back to "default".
 *
 *
 * @param {function} setterFunct - useState setter funct to set animation state
 * @param {integer} delay - delay of setterFunct being called
 * @param {funct} callback - function to call upon completion
 * @param {array} args - array of arguments for callback function
 */

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  // after delay set back to "default"
  return setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
