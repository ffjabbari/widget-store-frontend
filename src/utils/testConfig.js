/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'The only remote you will need',
    description: 'Lorem ipsum dolor sit, placeat.',
    discount: '0.15',
    image_url:
      'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    quantity: 1,
    rating: null,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: 'Apt 5',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 1,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };

  // mock return from /order/:id
  testData.orderGet = {
    customer: 1,
    customer_info: {
      first_name: 'Test',
      last_name: 'User',
      email: 'foo@gmail.com',
      address: '123 Main St',
      address_line2: '',
      city: 'Big City',
      state: 'TX',
      postal_code: '39488',
      phone_number: '(555) 555-5555',
    },

    discount: null,
    items: [
      { 0: { order_id: 2, product_id: 6, quantity: 1 } },
      { 1: { order_id: 2, product_id: 8, quantity: 1 } },
    ],

    order_date: '2021-02-05T08:00:00.000Z',
    order_id: 2,
    shipping_address: null,
    shipping_cost: '12.00',
    shipping_method: {
      details: { cost: '12.00', name: 'USPS ground shipping (3-7 days)' },
      shipping_method: 'usps_ground',
    },

    status: 'Pending',
    subtotal: '784.16',
    tax: '66.65',
    total: '862.81',
    total_items_quantity: 2,
    tracking_number: null,
  };

  TEST_DATA.shippingMethods = {
    usps_ground: {
      cost: '12.00',
      name: 'USPS ground shipping (3-7 days)',
    },
    usps_priority: {
      cost: '16.00',
      name: 'USPS priority shipping (2-4 days)',
    },
    usps_next_day: {
      cost: '25.00',
      name: 'USPS next day delivery (1 day)',
    },
    ups_ground: {
      cost: '16.00',
      name: 'UPS ground shipping (3-6 days)',
    },
    ups_2nd_day: {
      cost: '20.00',
      name: 'UPS second day delivery (2 days)',
    },
    ups_overnight: {
      cost: '28.00',
      name: 'UPS overnight delivery (1 day)',
    },
  };
};
