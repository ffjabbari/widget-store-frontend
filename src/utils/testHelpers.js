import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { testStore } from '../redux/store/reduxStore';

export const renderWithStore = component =>
  render(
    <Provider store={testStore}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>
  );
