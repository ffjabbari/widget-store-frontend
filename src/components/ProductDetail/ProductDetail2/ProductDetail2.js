/** Product Detail Modal Type 2
 *
 * Shows product details on a custom animated centered modal.
 */

import { useParams, useHistory, useLocation } from 'react-router-dom';
import React, { useState, useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AnimatePresence, motion } from 'framer-motion';

// local imports
import APIRequest from '../../../hooks/apiHook';
import { getPathRoot } from '../../../utils/helpers';
import toggleDir from '../../../redux/actions/animation/toggleDir';
import ScreenBackground from '../ModalBackground';
import PDModal from '../PDModalCard';
import '../ProductDetail.css';

const ProductDetail2 = () => {
  const { id } = useParams();
  const history = useHistory();
  const dispatch = useDispatch();
  const location = useLocation();
  const pathRoot = getPathRoot(location.pathname);
  const [modal, setModal] = useState(true);
  const slideoutDir = useSelector(state => state.animation.flipFlop);
  const body = useRef(document.body);

  useEffect(() => {
    // toggle animation direction
    dispatch(toggleDir());
    // add 'modal-open' class to <body> to remove scrolling
    const copy = body.current;
    copy.classList.add('modal-open');

    return () => copy.classList.remove('modal-open');
  }, [dispatch]);

  // close modal function
  const handleClose = e => {
    // only allow elements with class of clode-modal to close modal
    if (!e.target.classList.contains('close-modal')) return;
    setModal(false);
    setTimeout(() => history.push(pathRoot), 500);
  };

  // load product data
  const { loading, error, response } = APIRequest(`/products/${id}`, 'get');

  // if no product found go to path root i.e. /shop, /new, /deals
  if (error) {
    history.replace(pathRoot);
    return null;
  }
  if (loading) return null;

  const product = response.product;

  return (
    <AnimatePresence>
      {modal && (
        <motion.div
          initial={{ opacity: 0 }}
          animate={{ opacity: [0, 0.94, 0.95, 0.98, 1] }}
          exit={{ opacity: [1, 0.2, 0] }}
          transition={{ type: 'linear', ease: 'easeOut', delay: 0.1 }}
        >
          <ScreenBackground onClick={handleClose} className="close-modal">
            <motion.div
              initial={{
                x: 400 * (slideoutDir ? 1 : -1),
                y: -200,
                scale: 0.35,
                skewX: 9 * (slideoutDir ? 1 : -1) * -1,
                rotateX: '60deg',
                opacity: 0,
              }}
              animate={{
                x: 0,
                y: 0,
                scale: 1,
                skewX: 0,
                rotateX: '0deg',
                opacity: 1,
              }}
              exit={{
                x: 1100 * (slideoutDir ? 1 : -1) * -1,
                y: 0,
                scale: 0.2,
                skewX: 0,
                rotateX: '0deg',
                opacity: 0,
              }}
              transition={{
                duration: 0.8,
                type: 'spring',
                ease: 'easeIn',
                delay: 0.1,
              }}
            >
              <PDModal handleClose={handleClose} product={product} />
            </motion.div>
          </ScreenBackground>
        </motion.div>
      )}
    </AnimatePresence>
  );
};

export default ProductDetail2;
