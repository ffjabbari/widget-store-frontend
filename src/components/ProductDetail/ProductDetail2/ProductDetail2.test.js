/** ProductDetail2 tests */
import { act, fireEvent } from '@testing-library/react';

// local imports
import ProductDetail2 from './ProductDetail2';
import resetCart from '../../../redux/actions/cart/resetCart';
import { renderWithStore } from '../../../utils/testHelpers';
import { testStore } from '../../../redux/store/reduxStore';
import { TEST_DATA, populateTestDataHook } from '../../../utils/testConfig';

const axios = require('axios');
jest.mock('axios');

// mock useHistory
let mockUseHistoryReplace = jest.fn();
let mockUseHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    replace: mockUseHistoryReplace,
    push: mockUseHistoryPush,
  }),
}));

beforeAll(() => {
  populateTestDataHook(TEST_DATA);
});

beforeEach(() => {
  axios.mockResolvedValue({
    data: {
      product: TEST_DATA.product,
    },
  });
});

afterEach(() => {
  jest.resetAllMocks();
  testStore.dispatch(resetCart());
});

test('renders ProductDetail2', async () => {
  await act(async () => {
    renderWithStore(<ProductDetail2 />);
  });
});

test('ProductDetail2 snapshot', async () => {
  let asFragment;
  await act(async () => {
    const resp = renderWithStore(<ProductDetail2 />);
    asFragment = resp.asFragment;
  });
  expect(asFragment()).toMatchSnapshot();
});

test('Quantity input in document', async () => {
  let getByLabelText;
  await act(async () => {
    const resp = renderWithStore(<ProductDetail2 />);
    getByLabelText = resp.getByLabelText;
  });
  // test quantity input present with value of '1'
  const quantityInput = getByLabelText('Quantity');
  expect(quantityInput).toBeInTheDocument();
  expect(quantityInput.value).toBe('1');
});

test('closing modal ok', async () => {
  let getByRole;
  await act(async () => {
    const resp = renderWithStore(<ProductDetail2 />);
    getByRole = resp.getByRole;
  });
  const cancelButton = getByRole('button', { name: /cancel/i });
  await act(async () => {
    fireEvent.click(cancelButton);
    // await delay to allow animation then page url change
    await new Promise(res => setTimeout(res, 500));
  });

  expect(mockUseHistoryPush.mock.calls.length).toBe(1);
  expect(mockUseHistoryPush.mock.calls[0][0]).toBe('/');
});

test("clicking in modal doesn't close modal", async () => {
  let getByLabelText;
  await act(async () => {
    const resp = renderWithStore(<ProductDetail2 />);
    getByLabelText = resp.getByLabelText;
  });
  const quantityInput = getByLabelText(/quantity/i);
  await act(async () => {
    fireEvent.click(quantityInput);
    // await delay to allow animation then page url change
    await new Promise(res => setTimeout(res, 500));
  });

  expect(mockUseHistoryPush.mock.calls.length).toBe(0);
});

test('API error returns null', async () => {
  axios.mockImplementation(() => {
    throw new Error();
  });
  let queryByRole;
  await act(async () => {
    const resp = renderWithStore(<ProductDetail2 />);
    queryByRole = resp.queryByRole;
  });
  const cancelButton = queryByRole('button', { name: /cancel/i });
  expect(cancelButton).not.toBeInTheDocument();
  expect(mockUseHistoryReplace.mock.calls.length).toBe(1);
  expect(mockUseHistoryReplace.mock.calls[0][0]).toBe('/');
});
