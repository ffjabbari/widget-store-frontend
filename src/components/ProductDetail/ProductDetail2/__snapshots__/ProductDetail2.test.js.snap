// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail2 snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ itNZRq fadeIn close-modal"
  >
    <div
      style="opacity: 0; transform: translateX(400px) translateY(-200px) scale(0.01) rotateX(-60deg) skewX(-9deg) translateZ(0);"
    >
      <div
        class="sc-gsTCUz ldfCwt ProductDetail"
        color="var(--dark)"
        data-testid="ModalCard"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
            <div>
              Samsung XG-900 55 inch LCD TV
            </div>
          </h5>
          <button
            class="close-modal close"
          >
            Ã—
          </button>
        </div>
        <div
          class="p-0 row modal-body"
        >
          <div
            class="col-12 col-lg-6"
          >
            <img
              alt="Samsung XG-900 55 inch LCD TV"
              class="ProductDetail-img img-fluid"
              src="https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg"
            />
          </div>
          <div
            class="col-12 col-lg-6"
          >
            <div
              class="details p-2"
            >
              <h5
                class="mb-4 text-center pt-4"
              >
                Latest generation smart televison 2021
              </h5>
              <p>
                Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.
              </p>
              <div
                class="text-right pb-3 pr-5 mr-5"
              >
                <span
                  class="sc-dlfnbm kJMwKE"
                >
                  <span
                    class="sc-hKgILt gCMyNV price"
                  >
                    $400.40
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <form>
            <label
              for="quantity"
            >
              Quantity
            </label>
            <input
              id="quantity"
              min="1"
              name="quantity"
              placeholder="1"
              type="number"
              value="1"
            />
            <button
              class="btn btn-primary"
            >
              Add To Cart
            </button>
          </form>
          <div
            style="--btnAnimColor: white;"
          >
            <button
              aria-label="go to cart"
              class="default btn btn-secondary"
              id="cart-button"
              type="button"
            >
              <span
                class="CartIconBadge position-relative large"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-shopping-cart fa-w-18 fa-lg "
                  data-icon="shopping-cart"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 576 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M528.12 301.319l47.273-208C578.806 78.301 567.391 64 551.99 64H159.208l-9.166-44.81C147.758 8.021 137.93 0 126.529 0H24C10.745 0 0 10.745 0 24v16c0 13.255 10.745 24 24 24h69.883l70.248 343.435C147.325 417.1 136 435.222 136 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-15.674-6.447-29.835-16.824-40h209.647C430.447 426.165 424 440.326 424 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-22.172-12.888-41.332-31.579-50.405l5.517-24.276c3.413-15.018-8.002-29.319-23.403-29.319H218.117l-6.545-32h293.145c11.206 0 20.92-7.754 23.403-18.681z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </button>
          </div>
          <button
            class="close-modal btn btn-secondary"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
